Use asynchronous repo

thedaniel opened the pull request:
tree-view is the best candidate for exercising the new async repo class we’re working on in https://github.com/atom/atom/pull/9213

This depends on that branch of atom/atom. The plan is to merge the async repo after it proves sufficient for tree-view’s consumption, then merge this.

thedaniel commented:
Hm, I only have 2 (seemingly unrelated) failures on my machine, but there are a bunch of :red_circle: on travis. In any case this is starting to come together nicely.

thedaniel commented:
Of course, it's never as simple as 'just make git decorations async' - this refactor ended up touching a bunch of parts of the tree-view (which is fine, more async is better, right?) but it's been a slog. I'm hoping most of these remaining red dots are just specs that need to be updated to wait on asynchronous directory expansion, not bugs or races w/ the async git repo or within this package's views.

![screen shot 2015-12-24 at 5 49 42 pm](https://cloud.githubusercontent.com/assets/1476/11998897/f71cb6ce-aa66-11e5-8c4f-81c748b36f25.png)

thedaniel commented:
This is green on my machine :tada: with this change to atom - https://github.com/atom/atom/pull/10352 though it seems that PR needs a little more attention before it can merge.

<img width="627" alt="screen shot 2016-01-09 at 12 45 13 pm" src="https://cloud.githubusercontent.com/assets/1476/12218244/dc1b4f2e-b6ce-11e5-8613-26a3681be605.png">

The specs are deeply littered with waitsfor/runs grossness, but it seemed to be the most direct route to getting them green with this big refactor. In the future we should consider breaking up the spec sections into separate files and porting them to ES6 one by one.

Also there are a few opportunities to asyncify this further but given the original intention was to just start using the async git stuff, i figured more refactoring should be done in a separate PR rather than mixing it in here.

cc @atom/feedback - this isn't quite ready to merge but let's have a review anyway :smile_cat:

joshaber commented on lib/tree-view.coffee:
```diff
@@ -688,6 +728,8 @@ class TreeView extends View
 
       if repo = repoForPath(newPath)
+        # The return values of this are thrown away, they are called for emiiter
+        # side effects only.
```
`GitRepositoryAsync` has `.refreshStatusForPath` which does this, just with a clearer name.

joshaber commented:
:metal: Just :one: :memo:

thedaniel commented:
cc: @atom/feedback @atom/core 

Items for review / notable changes:

`tree-view.coffee`

- See big comment block at top of `initialize()` explaining use of a mutation observer that makes me feel not great.
- I serialized promises using Array.reduce in `initialize()` as we needed to expand recursively until we find (or don't find) a view for the path that's supposed to be selected on open.

`directory-view.coffee`

- I refactored the synchronous `expand()` to use a helper method called `expandAsync()` rather than doing the promise chaining inline, and realized that `expandAsync()` inside `waitsForPromise` is nice in tests. The old `expand()` returned false when done, and I kept that even though it is no longer sync, in case there were callers out there expecting that return value. Not sure if this will cause problems or if I should change the return value to a promise there as well.

`directory.coffee`

- `updateStatus()` used a lot of sync methods and got a lot more complicated using the async repo. It could use a sanity check.
- `getEntries()` really blew up in complexity. It was a very simple "loop over files and call sync status checks on each one" and I ultimately chose to build an array of promises checking statuses and then resolving them with Promise.all and sorting the last step of the promise chain. It seems fine, but the method doubled in size so there might be a clearer way to do it.
- I just changed `reload()` from sync to async, assuming that since tree-view doesn't really have a public API #itsfine.
- same with `expand()`

`file.coffee`

- `updateStatus()` became a 3-promise-chain async method with a bunch of conditionals, due to use of async staus checking methods

`tree-view-spec.coffee`

- So much ugly and repetitive conversion waitsFor and waitsForPromise :/ . In a dream world I'd love to break this up into 10 different spec files and rewrite them in ES6 using async/await but that's out of scope esp since it seems we're going to need to rewrite tree-view anyway to provide some kind of API to package developers.

nathansobo commented on lib/tree-view.coffee:
```diff
@@ -53,7 +53,34 @@ class TreeView extends View
     @selectEntry(@roots[0])
 
-    @selectEntryForPath(state.selectedPath) if state.selectedPath
+    # Now that directory expansion is done asyncly combining 2 different
```
:grimacing: This is why it would be really fantastic to have a view model.

thedaniel replied:
I wish we could figure out a way to rewrite tree-view incrementally without making it a giant project...

nathansobo commented on lib/tree-view.coffee:
```diff
@@ -301,15 +328,29 @@ class TreeView extends View
+    # this function) iterate over lists of values - generally we can just use
+    # Promise.all in those cases, but this one works its way sequentially through
+    # the list, so we use Array::reduce to serialize the promises.
+    activePathComponents.reduce reducer, Promise.resolve(rootPath)
```
Interesting use of `reduce` for sure. More wishful thinking, would be elegantly expressed in Babel with a for loop wrapping `await`. Oh well.

nathansobo commented on lib/file.coffee:
```diff
@@ -35,32 +35,33 @@ class File
-    else
-      status = repo.getCachedPathStatus(@path)
-      if repo.isStatusModified(status)
+    repo?.isPathIgnored(@path).then (isIgnored) =>
```
This all makes sense to me.

nathansobo commented on lib/directory.coffee:
```diff
@@ -30,8 +32,6 @@ class Directory
     @status = null
     @entries = {}
 
-    @submodule = repoForPath(@path)?.isSubmodule(@path)
```
Are we eliminating the `submodule` distinction with the new repository API or has it just moved elsewhere?

nathansobo commented on lib/directory.coffee:
```diff
@@ -163,34 +165,63 @@ class Directory
+      # throw this in a closure so we don't have scoping issues
+      # async/await would make this more pleasant when this is
+      # inevitably ported to ES6
+      f = =>
```
Would Coffee's `do` syntax help here at all for an anonymous immediately-invoked function?

maxbrunsfeld replied:
Since you're building up a corresponding array of promises and need a closure for scoping, it might make sense to use `Array.prototype.map`.

thedaniel replied:
Good call - `do` would be cleaner, but `map` is the right thing to do here, I think, I prefer using natural JS stuff over CS sugar unless it's really messy to do so.

nathansobo commented on lib/directory.coffee:
```diff
@@ -163,34 +165,63 @@ class Directory
+            if @entries.hasOwnProperty(localName)
+              # push a placeholder since this entry already exists but this helps
+              # track the insertion index for the created views
+              return [localName, 'directory']
```
I'm a bit lost here and might be confused, but could you just insert the existing directory object itself rather than the placeholder?

thedaniel replied:
I'm not sure. I wondered the same thing (it's the same way in the sync code) but I didn't fully grok why the placeholder was used at the time and just kept the existing behavior to avoid more breakage. Now that things are green again it's a better time to mess with it.

thedaniel replied:
Naïvely swapping in `@entries[localName]` lights the specs on fire - I'll spend a small amount of time digging into it before committing to the cargo cult

thedaniel replied:
Interestingly passing the `File` entry down is fine, it just causes problems when the `Directory` is passed down instead of `'directory'`. I wonder if this has to do with async expanding of the tree...

nathansobo commented on lib/directory.coffee:
```diff
@@ -163,34 +165,63 @@ class Directory
+            if @entries.hasOwnProperty(localName)
+              # push a placeholder since this entry already exists but this helps
+              # track the insertion index for the created views
+              return [localName, 'file']
```
Same here... could you use the object from the `@entries` map instead of a string?

nathansobo commented on lib/directory-view.coffee:
```diff
@@ -106,18 +113,41 @@ class DirectoryView extends HTMLElement
-      for entry in @entries.children when entry instanceof DirectoryView
-        entry.expand(true)
-
+      children = (entry for entry in @entries.children when entry instanceof DirectoryView)
```
I'm not understanding why you're special casing on the existence of `firstEntry`... Is it to check for the completion the view's directory's expansion? If so could you just always wait on the promise, which might already be resolved and do work immediately?

nathansobo commented:
Okay, I looked at everything. It seems as reasonable as it can be for the fact that this code is woefully in need of a view model layer to separate the async complexity from DOM concerns. Asked a couple questions and suggested a couple things.

maxbrunsfeld commented on lib/tree-view.coffee:
```diff
@@ -53,7 +53,34 @@ class TreeView extends View
+    #    with the new async git repo interaction. However, we also use event
+    #    subscription with Directory::onDidAddEntries to build up a tree view as
+    #    well, meaning we can't just tag on to the end of the Directory::expand()
+    #    promise chain :/
```
Naively, I would think we could have `updateRoots` return a promise that resolves after all of the directories are expanded, and do something like

```coffee
@updateRoots(state.directoryExpansionStates).then =>
  if state.selectedPath
    @selectEntryForPath(state.selectedPath)
  else
    @selectEntry(@roots[0])
```

I assume I'm missing something; how does the `Directory::onDidAddEntries` code path relate to this?

alexandernst commented:
Almost 2 months without any work/news on this. Is this dead?

thedaniel commented:
@alexandernst not dead, just got set aside while we did some other work. The async repo support only just landed in stable a week ago so we'll need to pick this back up again soon.

50Wliu commented:
I don't believe we're going with the route that this PR is taking, considering nodegit is no more in Atom core.
